### DISCUSSION TITLE
Is it possible to replace field accesses?

### DISCUSSION THREAD

**Discussion by Azn9** [raphw/byte-buddy#1514] [Q&A] [answered]
I have a class that looks like this:
```Java
public class Foo {

    private String foo;
    
    public String getFoo() {
        return this.foo;
    }
    
    public void doSomething() {
        String temp = this.foo;

        // ...
    }
}
```

Is there any way to modify this class so that the resulting "fake" class would act like the following?
```Java
//...
    public void doSomething() {
        String temp = this.getFoo(); // <--- replace "this.foo" by its getter
//...
}
```

I'm already replacing the getter, but sometimes in the source code we use "direct" access to the fields, that would not give the same value than the getter.

**Comment by raphw** [comment, marked as answer]
Yes, you are looking for MemberSubstitution. You'd replace a field read with a method invocation.

**Reply by Azn9** [reply]
I tried with something like the following, but I got the error "class redefinition failed: attempted to change the schema (add/remove fields)", do you see any mistake that I made? I don't find it :'(
It's my first time using bytebuddy and I'm a bit lost with all the possibilities this library offer 😅 

```Java
DynamicType.Builder<T> builder = LiveObjectUtils.BYTE_BUDDY
        .redefine(clazz)
        .method(this::isBlockingGetter).intercept(MethodDelegation.to(BlockingGetterInterceptor.class))
        .method(this::isReactiveGetter).intercept(MethodDelegation.to(ReactiveGetterInterceptor.class))
        .method(this::isReactiveSetter).intercept(MethodDelegation.to(ReactiveSetterInterceptor.class))
        .method(this::isBlockingSetter).intercept(MethodDelegation.to(BlockingSetterInterceptor.class));

for (Field declaredField : clazz.getDeclaredFields()) {
    builder = builder.visit(MemberSubstitution.relaxed()
                    .field(ElementMatchers.named(declaredField.getName()))
                    .onRead()
                    .replaceWithMethod(new GetterMatcher(declaredField.getName(), declaredField.getType()))
                    .on(ElementMatchers.definedMethod(ElementMatchers.any()))
            );
}

try (DynamicType.Unloaded<T> unloadedT = builder.make()) {
    Class<? extends T> dynamicType = unloadedT
            .load(this.getClass().getClassLoader(), ClassReloadingStrategy.fromInstalledAgent())//ClassLoadingStrategy.Default.INJECTION)
            .getLoaded();

// ...
```

**Reply by Azn9** [reply]
Okay, if I remove all my .method().intercept(), it works (but I get a stackoverflow as it also replaces the one in the getter, making it call itself)
Does the interception adds methods to the class?

**Reply by raphw** [reply]
Interception adds metgods containing the original code. You'd need to use Advice then and use visit to decorate methods.

**Reply by Azn9** [reply]
I finally managed to get it work as I expected, thank you very much for your help!

I just have one last question, I used the following code to skip the method execution and replace the return value, but I was wondering if this is the correct way to do it?
```
@Advice.OnMethodEnter(skipOn = Boolean.class)
static Object enter() {
    return Boolean.TRUE;
}

@Advice.OnMethodExit(onThrowable = Throwable.class)
static void exit(@Advice.Return(readOnly = false, typing = Assigner.Typing.DYNAMIC) Object returnValue) {
    returnValue = "Returned from advice :o";
}
```

**Comment by raphw** [comment]
That's one way. You can also set Advice.OnNonDefaultValue.class and return a primitive, but I doubt it makes a performance difference.

**Reply by Azn9** [reply]
Thank you very much for all the help!

**Comment by manoelcampos** [comment]
I have a similar issue, but I need to replace field access inside a byte buddy Plugin implementation.
Consider I have these classes:

```java

@Entity
class MyEntity{
   public int id;
   public String name;

  // getters and setters
}

class Client {
    void someMethod(){
        var entity = new MyEntity();
        entity.id = 1; // must be replaced by entity.setId(1)
        int id = entity.id; // must be replaced by entity.getId()
        System.out.println("Id: " + id); 
    }
}
```

My plugin class is this one:

```java
public class EntityAccessorInstrumentationPlugin implements Plugin {
    @Override
    public DynamicType.Builder<?> apply(
        final DynamicType.Builder<?> builder,
        final TypeDescription typeDescription,
        final ClassFileLocator classFileLocator)
    {
        final Junction<FieldDescription> isInstanceField = isPublic().and(not(isStatic()));

        // Intends to replace public instance fields reads by the respective getter call
        final var methodWithFieldReadVisitor = MemberSubstitution.relaxed()
                                                       .field(isInstanceField)
                                                       .onRead()
                                                       .replaceWithMethod(ElementMatchers.isGetter())
                                                       .on(ElementMatchers.isMethod());

        // Intends to replace public instance fields writes by the respective setter call
        final var methodWithFieldWriteVisitor = MemberSubstitution.relaxed()
                                                        .field(isInstanceField)
                                                        .onWrite()
                                                        .replaceWithMethod(ElementMatchers.isSetter())
                                                        .on(ElementMatchers.isMethod());

        return builder.visit(methodWithFieldReadVisitor).visit(methodWithFieldWriteVisitor);
    }
}
```

The issue is that, for instance, the methodWithFieldReadVisitor needs to find the getter matching the field that is being read and replace by the getter call, but the `ElementMatchers.isGetter()` doesn't try to find the getter for the accessed field. Instead, it tries to find all getters.

**Reply by raphw** [reply]
You can implement your own matcher, wouldn't that work?

**Reply by manoelcampos** [reply]
I need to implement a Method Matcher, but the matcher just receives a MethodDescriptor. How can I know the name of the field to look for a matching getter/setter? Implementing the matcher is easy, but without the field name it's not possible:

```java
    class GetterMatcher extends ElementMatcher.Junction.AbstractBase<MethodDescription> {
        @Override
        public boolean matches(MethodDescription methodDescription) {
            // TODO How can I get the field being accessed so that I can lookup for an accessor method?
            return true;
        }
    }

```

**Comment by raphw** [comment]
I see, you'd need your own substitution. From it, you can investigate the intercepted element and simply delegate to the desired substitution that already ships with Byte Buddy, or implement it yourself and use some of the elements that are preshipped.

**Reply by manoelcampos** [reply]
Thanks for the direction
